src/back/glsl_common.rs

```diff
@@ -0,0 +1,704 @@
+                            ScalarKind::Uint => "u",
+                            ScalarKind::Float => match width {
+                                4 => "",
+                                8 => "d",
+                                _ =>
+                                    return Err(Error::Custom(format!(
+                                        "Cannot build float of width {}",
+                                        width
+                                    ))),
+                            },
+                            ScalarKind::Bool => "b",
+                        },
+                        size as u8,
+                    ),
+                    TypeInner::Matrix {
+                        columns,
+                        rows,
+                        kind,
+                        width,
+                    } => format!(
+                        "{}mat{}x{}",
+                        match kind {
+                            ScalarKind::Sint => "i",
+                            ScalarKind::Uint => "u",
+                            ScalarKind::Float => match width {
+                                4 => "",
+                                8 => "d",
+                                _ =>
+                                    return Err(Error::Custom(format!(
+                                        "Cannot build float of width {}",
+                                        width
+                                    ))),
+                            },
+                            ScalarKind::Bool => "b",
+                        },
+                        columns as u8,
+                        rows as u8,
+                    ),
+                    TypeInner::Array { .. } => ty.write_glsl(builder.types, builder.structs)?,
+                    TypeInner::Struct { .. } => builder.structs.get(ty).unwrap().clone().0,
+                    _ =>
+                        return Err(Error::Custom(format!(
+                            "Cannot compose type {}",
+                            ty.write_glsl(builder.types, builder.structs)?
+                        ))),
+                },
+                components
+                    .iter()
+                    .map(|arg| builder.expressions[*arg].write_glsl(module, builder))
+                    .collect::<Result<Vec<_>, _>>()?
+                    .join(","),
+            ),
+            Expression::FunctionParameter(pos) => builder.args.get(&pos).unwrap().to_string(),
+            Expression::GlobalVariable(handle) => builder.globals.get(&handle).unwrap().clone(),
+            Expression::LocalVariable(handle) => {
+                builder.locals_lookup.get(&handle).unwrap().clone()
+            }
+            Expression::Load { pointer } => todo!(),
+            Expression::ImageSample {
+                image,
+                sampler,
+                coordinate,
+                depth_ref,
+            } => {
+                let image_ty = crate::proc::Typifier::new().resolve(
```

Comment:
we should retain the typifier instead of creating a new one: it caches types as it traverses the module